File: RoboGraph/Source/RoboGraph/GraphCore/RGAdjacencyMatrix.cpp
// Fill out your copyright notice in the Description page of Project Settings.

#include "RGAdjacencyMatrix.h"

RGGraphCore::RGAdjacencyMatrix::RGAdjacencyMatrix(int InSize)
{
	Matrix = new float[InSize * InSize]; // Light weight solution to implement 2d arrays
	Size = InSize;
}

void RGGraphCore::RGAdjacencyMatrix::AddDirectedEdge(int From, int To, float Weight)
{
	Matrix[From * Size + To] = Weight;
}

void RGGraphCore::RGAdjacencyMatrix::AddUndirectedEdge(int V1, int V2, float Weight)
{
	Matrix[V1 * Size + V2] = Weight;
	Matrix[V2 * Size + V1] = Weight;
}

float RGGraphCore::RGAdjacencyMatrix::GetEdgeWeight(int X, int Y)
{
	return Matrix[X * Size + Y];
}

TArray<int> RGGraphCore::RGAdjacencyMatrix::GetAdjacencyList(int SourceIndex)
{
	TArray<int> AdjacencyList;
	for (int ListIndex = 0; ListIndex < Size; ListIndex++)
	{
		if (Matrix[SourceIndex * Size + ListIndex] != 0)
		{
			AdjacencyList.Add(ListIndex);
		}
	}

	return AdjacencyList;
}

File: RoboGraph/Assets/Scripts/RGAdjacencyMatrix.cs
using System.Collections.Generic;

namespace RGGraphCore
{
    public class RGAdjacencyMatrix
    {
        private float[,] _matrix;
        private int _size;
        public int Size { get { return _size; } }

        public RGAdjacencyMatrix(int size)
        {
            _matrix = new float[size, size];
            _size = size;
        }

        public void AddDirectedEdge(int from, int to, float weight)
        {
            _matrix[from, to] = weight;
        }

        public void AddUndirectedEdge(int V1, int V2, float weight)
        {
            _matrix[V1, V2] = weight;
            _matrix[V2, V1] = weight;
        }

        public float GetEdgeWeight(int x, int y)
        {
            return _matrix[x, y];
        }

        public List<int> GetAdjacencyList(int sourceIndex)
        {
            List<int> adjacencyList = new List<int>();
            for (int i = 0; i < _size; i++)
            {
                if (_matrix[sourceIndex, i] != 0)
                {
                    adjacencyList.Add(i);
                }
            }

            return adjacencyList;
        }
    }
}